### DISCUSSION TITLE
Can't get UUID to work with Postgres

### DISCUSSION THREAD

**Discussion by harrismcc** [diesel-rs/diesel#3623] [Q&A] [answered]
I'm having trouble getting Diesel to play nice with UUIDs. I have an `id` column in my table that is of type `UUID`, but when I try to use it, no matter what I try, I get the error message below.

I'm thinking it's because I don't have the right type for `id` for my `TradeGood` struct in `models.rs`. I know from some other threads here that I'm not supposed to use the same `Uuid` type from `schema.rs` (I tried anyway, no dice). I tried switching the types in both `schema.rs` and `models.rs` to `String` and that makes the error go away, so I'm sure that it's this Uuid type.

Any ideas for what I'm missing here? I think I just need the right type to plug in, but I can't find what that would be...
(P.S. If it wasn't obvious, I'm new to Rust so beginner-friendly explanation may be needed. Thanks!)

### Error Message
```
error[E0277]: the trait bound `(diesel::sql_types::Uuid, Text, Integer, schema::sql_types::SupplyType, Integer, Integer): load_dsl::private::CompatibleType<TradeGood, Pg>` is not satisfied
    --> src/main.rs:25:28
     |
25   |         .load::<TradeGood>(connection)
     |          ----              ^^^^^^^^^^ the trait `load_dsl::private::CompatibleType<TradeGood, Pg>` is not implemented for `(diesel::sql_types::Uuid, Text, Integer, schema::sql_types::SupplyType, Integer, Integer)`
     |          |
     |          required by a bound introduced by this call
     |
     = help: the following other types implement trait `load_dsl::private::CompatibleType<U, DB>`:
               (ST0, ST1)
               (ST0, ST1, ST2)
               (ST0, ST1, ST2, ST3)
               (ST0, ST1, ST2, ST3, ST4)
               (ST0, ST1, ST2, ST3, ST4, ST5)
               (ST0, ST1, ST2, ST3, ST4, ST5, ST6)
               (ST0, ST1, ST2, ST3, ST4, ST5, ST6, ST7)
               (ST0, ST1, ST2, ST3, ST4, ST5, ST6, ST7, ST8)
             and 24 others
     = note: required for `SelectStatement<FromClause<table>, DefaultSelectClause<FromClause<table>>, NoDistinctClause, ..., ..., ...>` to implement `LoadQuery<'_, _, TradeGood>`
     = note: the full type name has been written to '/workspace/target/debug/deps/astravis_connector-fa3fe98ad1f8c86c.long-type-14754057602441959740.txt'
note: required by a bound in `diesel::RunQueryDsl::load`
    --> /usr/local/cargo/registry/src/github.com-1ecc6299db9ec823/diesel-2.0.4/src/query_dsl/mod.rs:1499:15
     |
1499 |         Self: LoadQuery<'query, Conn, U>,
     |               ^^^^^^^^^^^^^^^^^^^^^^^^^^ required by this bound in `RunQueryDsl::load`

```
### models.rs
```rust
use diesel::prelude::*;
use diesel_derive_enum;

#[derive(Debug, PartialEq, Clone, diesel_derive_enum::DbEnum)]
#[ExistingTypePath = "crate::schema::sql_types::SupplyType"]
pub enum SupplyType {
    SCARCE,
    LIMITED,
    MODERATE,
    ABUNDANT,
}

#[derive(Queryable, Clone)]
pub struct TradeGood {
    pub id: ??????, // <-- What do I put here?
    pub symbol: String,
    pub trade_volume: i32,
    pub supply: SupplyType,
    pub purchase_price: i32,
    pub sell_price: i32,
}

```
### schema.rs
```rust
// @generated automatically by Diesel CLI.

pub mod sql_types {
    #[derive(diesel::sql_types::SqlType)]
    #[diesel(postgres_type(name = "supply_type"))]
    pub struct SupplyType;
}

diesel::table! {
    use diesel::sql_types::*;
    use super::sql_types::SupplyType;

    market_trade_goods (id) {
        id -> Uuid,
        symbol -> Varchar,
        trade_volume -> Int4,
        supply -> SupplyType,
        purchase_price -> Int4,
        sell_price -> Int4,
    }
}
```

### Table SQL
```sql
-- Your SQL goes here
CREATE TABLE "market_trade_goods" (
  id UUID PRIMARY KEY,
  symbol VARCHAR NOT NULL,
  trade_volume INT NOT NULL,
  supply supply_type NOT NULL,
  purchase_price INT NOT NULL,
  sell_price INT NOT NULL
);
```

**Comment by weiznich** [comment, marked as answer]
Based on the provided code examples it's unclear what you actually tried to use as field type for your `id` field. 

Generally speaking: Diesel works with two distinct sets of types, rust side types and sql side types. The sql side types are used in the `table!` definition. They power the compile time check and they are essentially zero sized types. That means they cannot hold any values. Usually you shouldn't use them anywhere else than the `table!` macro. Rust side types are used in your model structs (Insertable, Queryable, AsChangeset) and as bind parameters in your directly constructed queries. 

There is an m:n mapping between rust side types and SQL side types, which means one rust type can map to several SQL types and one sql type can be mapped to several rust types. If you mix up that mapping you get error messages like that one.  (If you see such a error message it's in 99% of the cases such a mismatch). 

The documentation of the corresponding sql side types list all compatible rust side types. [Example for the sql uuid type](https://docs.diesel.rs/2.0.x/diesel/pg/sql_types/struct.Uuid.html). If you click on [`ToSql`](https://docs.diesel.rs/2.0.x/diesel/serialize/trait.ToSql.html#foreign-impls) or [`FromSql`](https://docs.diesel.rs/2.0.x/diesel/deserialize/trait.FromSql.html#foreign-impls) you get a list of all possible mappings. (If you open this pages on a locally build documentation version it will also include custom extensions like your `SupplyType`).

**Reply by harrismcc** [reply]
Thanks so much for this explainer. I think in my own debugging I got within inches of this final answer but never quite, so this was exactly what I needed to understand it.

For anyone else who might come across this, this is the code that worked for me in the end. I needed to use `uuid::Uuid` in `models.ts` and not `diesel::pg::sql_types::Uuid` which belongs only in `schema.rs`
### models.rs
```rust
use diesel::prelude::*;
use diesel_derive_enum;
use uuid::Uuid;

#[derive(Debug, PartialEq, Clone, diesel_derive_enum::DbEnum)]
#[ExistingTypePath = "crate::schema::sql_types::SupplyType"]
pub enum SupplyType {
    Scarce,
    Limited,
    Moderate,
    Abundant,
}

#[derive(Queryable, Clone)]
pub struct TradeGood {
    pub id: Uuid,
    pub symbol: String,
    pub trade_volume: i32,
    pub supply: SupplyType,
    pub purchase_price: i32,
    pub sell_price: i32,
}

```
